### PR TITLE
test: Add regression test on ATLAS sbottom public likelihood

### LINKED ISSUE #518: add public sbottom likelihood as test
# Description

once public we would want to add a wide range of public likelihoods into the CI as integration tests. Not breaking these likelihoods will be a priority of pyhf for future pyhf releases and probably should be an arbiter on whether the pyhf version in the spec needs to be updated for breaking changes

### PR DESCRIPTION
**Pull request by matthewfeickert**
# Description

Resolves #518 

Add a regression test that uses the observed and expected CLs values of the [ATLAS sbottom published likelihood](https://www.hepdata.net/record/ins1748602)

# Checklist Before Requesting Reviewer

- [x] Tests are passing
- [x] "WIP" removed from the title of the pull request
- [x] Selected an Assignee for the PR to be responsible for the log summary

# Before Merging

For the PR Assignees:

- [x] Summarize commit messages into a comprehensive review of the PR

```
* Add regression test on ATLAS sbottom public likelihood
   - c.f. https://www.hepdata.net/record/ins1748602
```

### CODE REVIEW

**Comment by matthewfeickert** [comment]
For the Region C test I had to raise the tolerance to `1e-4` as Python 2.7 on Travis CI was failing for one of the CLs_exp points with the tolerance at `1e-5`. However, all the Python 3 instances on GitHub actions passed this fine, so once we drop Python 2.7 in Issue #469 then we can set the tolerance back to `1e-5` again.

**Comment by matthewfeickert** [comment]
If there are more intelligent or more interesting points to be testing let me know and I will happily add them.

**Comment by kratsg** [comment]
> For the Region C test I had to raise the tolerance to `1e-4` as Python 2.7 on Travis CI was failing for one of the CLs_exp points with the tolerance at `1e-5`. However, all the Python 3 instances on GitHub actions passed this fine, so once we drop Python 2.7 in Issue #469 then we can set the tolerance back to `1e-5` again.

Just drop python 2.7. It's not even worth getting it to pass. We made all of the validations for the pub note on python3 where possible.

Where are you getting those numbers from btw? For RegionC, I have

```
$ cat data/outputs/results/pyhf/RegionC/pyhf_result_sbottom_1000_205_60.json 
{
    "CLs_exp": [
        0.8906470732412857,
        0.9280262743211622,
        0.9614288407343238,
        0.9857553063165135,
        0.9971958190844394
    ],
    "CLs_obs": 0.9424009499519606
}```

Additionally, with the way the tests are written, it's very hard to understand what results we're comparing because there's way too much modularization...

File: tests/test_regression.py [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+import os
+import pyhf
+import numpy as np
+
+
+@pytest.fixture(scope='module')
+def sbottom_likelihoods_download():
+    """Download the sbottom likelihoods tarball from HEPData"""
+    sbottom_HEPData_URL = "https://www.hepdata.net/record/resource/997020?view=true"
+    targz_filename = "sbottom_workspaces.tar.gz"
```
Comment (lukasheinrich) [inline tests/test_regression.py]: should we compare some download checksums here?

**Comment by matthewfeickert** [comment]
> Just drop python 2.7. It's not even worth getting it to pass. We made all of the validations for the pub note on python3 where possible.

I'll let you and @lukasheinrich discuss this here, as in the past when this has gotten brought up we've agreed based on his concerns to drop Python 2.7 at the end of the year after having cut a final release so that we avoid breaking people who are Python 2.7 bound for as long as possible.
 
> Where are you getting those numbers from btw? For RegionC, I have
> 
> ```
> $ cat data/outputs/results/pyhf/RegionC/pyhf_result_sbottom_1000_205_60.json 
> {
>     "CLs_exp": [
>         0.8906470732412857,
>         0.9280262743211622,
>         0.9614288407343238,
>         0.9857553063165135,
>         0.9971958190844394
>     ],
>     "CLs_obs": 0.9424009499519606
> }
> ``` 

```
$ curl -sL https://www.hepdata.net/record/resource/997020?view=true \
   | tar -O -xzv RegionC/BkgOnly.json \
   | pyhf cls --patch \
    <(curl -sL https://www.hepdata.net/record/resource/997020?view=true \
   | tar -O -xzv RegionC/patch.sbottom_1000_205_60.json)
RegionC/BkgOnly.json
RegionC/patch.sbottom_1000_205_60.json
{
    "CLs_exp": [
        0.8906506884676416,
        0.9280287127442725,
        0.9614301796189283,
        0.9857558128338463,
        0.9971959212073871
    ],
    "CLs_obs": 0.9424021663134358
}
```

> Additionally, with the way the tests are written, it's very hard to understand what results we're comparing because there's way too much modularization...

How so? Beyond the fixtures, which I think are reasonable, there are only two functions, and given the boilerplate nature of them I don't see how having them as functions is a problem. Can  you suggest a better naming scheme if things are confusing?

File: tests/test_regression.py (hunk 1 above)
Comment (matthewfeickert) [reply tests/test_regression.py]: This is a good idea.

**Comment by kratsg** [comment]
> How so? Beyond the fixtures, which I think are reasonable, there are only two functions, and given the boilerplate nature of them I don't see how having them as functions is a problem. Can you suggest a better naming scheme if things are confusing?

When you run the test, it only states "RegionA" -- it doesn't list the mass point that was run over. That should be part of the test name.

You should use my numbers for the RegionC -- since that's what the pub note was published on (which means we're regressing to the exact version of pyhf, but had to delay adding the test).

**Comment by matthewfeickert** [comment]
> When you run the test, it only states "RegionA" -- it doesn't list the mass point that was run over. That should be part of the test name.

That's fair. My thought was to keep the naming simple and then if multiple points were going to be tested change the name then, but you're right that it is better to be explicit up front now rather then have someone have to deal with name confusion or conflicts in the future. Let me know if you want changes to the modifications I've made.
 
> You should use my numbers for the RegionC -- since that's what the pub note was published on (which means we're regressing to the exact version of pyhf, but had to delay adding the test).

Sure. I've added them for Region C. Can you either give me these or add them for Region A and Region B?

File: tests/test_regression.py (hunk 1 above)
Comment (matthewfeickert) [reply tests/test_regression.py]: @lukasheinrich Done.

**Comment by lukasheinrich** [comment]
I think it woudl be sueful to add some points close to the contour e.g. 1400,1000 or so

**Comment by kratsg** [comment]
```
Lord Stark:~/sbottom-lhood-pubnote-code (triple_overlay)$ cat data/outputs/results/pyhf/RegionA/*1300_205_60*
{
    "CLs_exp": [
        0.09022509053507759,
        0.1937839194960632,
        0.38432344933992,
        0.6557757334303531,
        0.8910420971601081
    ],
    "CLs_obs": 0.24443627759085326
}Lord Stark:~/sbottom-lhood-pubnote-code (triple_overlay)$ cat data/outputs/results/pyhf/RegionA/*1400_950_60*
{
    "CLs_exp": [
        0.002644707461012826,
        0.013976754489151644,
        0.06497313811425813,
        0.23644505123524753,
        0.5744843501873754
    ],
    "CLs_obs": 0.021373283911064852
}Lord Stark:~/sbottom-lhood-pubnote-code (triple_overlay)$ cat data/outputs/results/pyhf/RegionA/*1500_850_60*
{
    "CLs_exp": [
        0.0059847029077065295,
        0.026103516126601122,
        0.10093985752614597,
        0.3101988586187604,
        0.6553686728646031
    ],
    "CLs_obs": 0.04536774062150508
}Lord Stark:~/sbottom-lhood-pubnote-code (triple_overlay)$ cat data/outputs/results/pyhf/RegionB/*1500_850_60*
{
    "CLs_exp": [
        0.9999682931142808,
        0.9999796235899027,
        0.9999893398756928,
        0.9999961574993774,
        0.999999261851925
    ],
    "CLs_obs": 0.9999893250416583
}Lord Stark:~/sbottom-lhood-pubnote-code (triple_overlay)$ cat data/outputs/results/pyhf/RegionC/*1000_205_60*
{
    "CLs_exp": [
        0.8906470732412857,
        0.9280262743211622,
        0.9614288407343238,
        0.9857553063165135,
        0.9971958190844394
    ],
    "CLs_obs": 0.9424009499519606
}
```

**Comment by matthewfeickert** [comment]
@kratsg Can you give me a different point for Region B than `1500_850_60`? Unless @lukasheinrich and you can agree on when to drop Python 2.7 support then the point you chose for Region B will require me to raise the tolerance from `1e-5` to `1e-4` for Python 2.7 on Travis CI.

**Comment by kratsg** [comment]
> raise the tolerance from `1e-5` to `1e-4`

You mean lower, but ok... try this one?

```
$ cat data/outputs/results/pyhf/RegionB/pyhf_result_sbottom_1400_550_60.json 
{
    "CLs_exp": [
        0.9338879894557114,
        0.9569045303300702,
        0.9771296335437559,
        0.9916370124133669,
        0.9983701133999316
    ],
    "CLs_obs": 0.9744675266677597
}
```

**Comment by matthewfeickert** [comment]
> You mean lower, but ok... try this one?

Yes. Sorry.

strike-that

Also that works now, so thanks.

**Comment by matthewfeickert** [comment]
@kratsg Can I also get output for Region C point `1600_850_60`?

**Comment by kratsg** [comment]
```
$ cat ~/sbottom-lhood-pubnote-code/data/outputs/results/pyhf/RegionC/*1600_850_60*
{
    "CLs_exp": [
        0.2955492909588046,
        0.4446885457298284,
        0.6371473864200973,
        0.8336149623750603,
        0.9585901381554178
    ],
    "CLs_obs": 0.711023707425625
}
```

**Comment by matthewfeickert** [comment]
@kratsg @lukasheinrich I think this is ready for review again now.

**Comment by matthewfeickert** [comment]
Thanks for the helpful reviews on this &mdash; it definitely needed them.

**Pull request merged**